generate/unicodegen/main.go

```diff
@@ -36,18 +38,101 @@ func visit(rt *unicode.RangeTable, fn func(rune)) {
 }
 
 func writeUnicode() error {
-	var b strings.Builder
-	visit(unicode.Po, func(r rune) {
-		if r != '"' {
-			b.WriteRune(r)
+	// rfc3987Ranges contains the ranges of valid code points specified by RFC 3987.
+	rfc3987Ranges := [][2]rune{
+		[2]rune{0xA0, 0xD7FF},
+		[2]rune{0xF900, 0xFDCF},
+		[2]rune{0xFDF0, 0xFFEF},
+		[2]rune{0x10000, 0x1FFFD},
+		[2]rune{0x20000, 0x2FFFD},
+		[2]rune{0x30000, 0x3FFFD},
+		[2]rune{0x40000, 0x4FFFD},
+		[2]rune{0x50000, 0x5FFFD},
+		[2]rune{0x60000, 0x6FFFD},
+		[2]rune{0x70000, 0x7FFFD},
+		[2]rune{0x80000, 0x8FFFD},
+		[2]rune{0x90000, 0x9FFFD},
+		[2]rune{0xA0000, 0xAFFFD},
+		[2]rune{0xB0000, 0xBFFFD},
+		[2]rune{0xC0000, 0xCFFFD},
+		[2]rune{0xD0000, 0xDFFFD},
+		[2]rune{0xE1000, 0xEFFFD},
+	}
+	removeRune := func(ranges [][2]rune, cp rune) [][2]rune {
+		for i, r := range ranges {
+			// Ranges are in ascending order. Skip any that precede `cp`,
+			// and bail out upon reaching one that follows `cp`.
+			if r[1] < cp {
+				continue
+			} else if cp < r[0] {
+				break
+			}
+
+			// `cp` is in this range and must be removed from it.
+			if cp == r[0] && cp == r[1] {
+				// Remove this single-element range.
+				return append(ranges[0:i], ranges[i+1:]...)
+			} else if cp == r[0] {
+				// Remove the first element of this range.
+				newRange := [2]rune{r[0] + 1, r[1]}
+				newTail := append([][2]rune{newRange}, ranges[i+1:]...)
+				return append(ranges[0:i], newTail...)
+			} else if cp == r[1] {
+				// Remove the last element of this range.
+				newRange := [2]rune{r[0], r[1] - 1}
+				newTail := append([][2]rune{newRange}, ranges[i+1:]...)
+				return append(ranges[0:i], newTail...)
+			} else {
+				// Split this range.
+				newTail := append(
+					[][2]rune{
+						[2]rune{r[0], cp - 1},
+						[2]rune{cp + 1, r[1]},
+					},
+					ranges[i+1:]...)
+				return append(ranges[0:i], newTail...)
+			}
 		}
+		return ranges
+	}
+
+	// sepFreeRanges excludes separators from rfc3987Ranges.
+	sepFreeRanges := append([][2]rune{}, rfc3987Ranges...)
+	visit(unicode.Z, func(cp rune) {
+		sepFreeRanges = removeRune(sepFreeRanges, cp)
```

Comment:
these are essentially triple `for` loops which call `append`, so I imagine it could get very slow if we're not careful. have you noticed `go generate` being much slower than before?

I'm not _that_ worried as long as unicodegen still finishes in much less than a second.

Reply:
For reference, `time go run ./generate/unicodegen` reports ~100ms for me.

Reply:
Sounds good. Just wanted to double check.